### ISSUE TITLE
[Feat] jpeg decompression using pytorch

### ISSUE THREAD

**Issue by narain1** [kornia/kornia#703]
## Pitch

> import kornia as K
> 
> img = open('image.jpeg')  # reading file content
> img = K.decompress_jpeg(img)

**Comment by edgarriba** [comment]
@narain1 any suggestion about how to start doing that ? Seems a nice feature to consider.

**Comment by narain1** [comment]
here is a good resource on jpeg decompression

[https://yasoob.me/posts/understanding-and-writing-jpeg-decoder-in-python/](url)

but the problem i still see here is we have to dis-assemble the file on the CPU and the we can move it as matrices to the target device

**Comment by edgarriba** [comment]
@narain1 feel free to draft a dummy approach first. We can think later how to optimise

**Comment by narain1** [comment]
[pytorch`s jpeg decoder](https://github.com/pytorch/vision/blob/main/torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp)

would it be of interest to add it here

or do you have any other additional functionalities in you can think which would be of better use?

**Comment by edgarriba** [comment]
@narain1 how do you propose to add ?

**Comment by narain1** [comment]
I don't have much of an Idea

pytorch`s  code looks much efficient as It has its own dispatch for both cpu and gpu.

So at this point you can guide better than me on how to proceed.

**Comment by edgarriba** [comment]
As a quick idea since tv API is still quite limited - we could externalize and combine also with opencv until we can host ourselves the implementation.

**Comment by lferraz** [comment]
Looks like TV is using nvjpeg. Shouldn't be so difficult to write the high-level api for a guy with experience using nvidia c++ libs. If we want to do it in tpus then probably we need to write all the code from scratch.

**Comment by cceyda** [comment]
as a side note torchvision nvjpeg support is still in beta.
I don't think I have ever seen TPU jpeg decoding, so that would be nice to have.

**Comment by edgarriba** [comment]
Ideally we might have to define a high-level API so that we can pass any device/backend and be agnostic  to the underlying library
```python
reader = ImageReader(format: str = ['png', 'jpg', 'tiff', 'hdr', ...], device = ['cpu', 'cuda', 'xla''])

img_t: torch.Tensor = reader('my_img.png')  # 1xCxHxW
```

**Comment by lferraz** [comment]
for video this [lib](https://github.com/NVIDIA/VideoProcessingFramework) is brutal from a pytorch perspective. There are things that need to be improved but in terms of functionality it works very well.

**Comment by lferraz** [comment]
I think TPU jpeg decoding could be doable if we put a bit of effort. But I do not know if it is something really useful (e.g. enough requests). On the other hand, we need TPUs to do it and they are not for free... so maybe we could talk with Google to get 100% free machines with TPUs.

**Comment by johnnv1** [comment]
@edgarriba just to confirm, does the `kornia-rs` do a "normal" decode and just pack it for tensor?

**Comment by edgarriba** [comment]
@johnnv1 right, `kornia-rs` implements Rust image library which has a wide range of decoder formats and specifically for jpeg we support turbojpeg via the `ImageDecoder` class.
* https://github.com/kornia/kornia-rs/blob/main/src/io.rs

More referring to the original question in this thread, i was checking the [original code](https://github.com/aguaviva/micro-jpeg-visualizer/blob/master/micro-jpeg-visualizer.py#L204) of the python jpeg decoder, buy i don't see where we could use torch compile to boost the decoding algorithm or even parallelize. Everything seems quite sequential here.
